ion-schema/src/model/constraints/ieee754_float.rs

```diff
@@ -0,0 +1,118 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::internal_traits::{
+    LoaderContext, ValidateInternal, ValidationContext, WriteAsIsl, WriteContext,
+};
+use crate::model::constraints::{ConstraintName, ReadConstraint};
+use crate::model::TypeDefinitionBuilder;
+use crate::result::{invalid_schema_error, IonSchemaError, IonSchemaResult};
+use crate::{IonSchemaElement, IslVersion, ViolationRecorder, ISL_1_0, ISL_2_0};
+use ion_rs::{Element, Value, ValueWriter};
+use std::ops::ControlFlow;
+
+impl ConstraintName for Ieee754Float {
+    const CONSTRAINT_NAME: &'static str = "ieee754_float";
+}
+
+/// Represents the `ieee754_float` constraint (ref.  [ISL 2.0]).
+///
+/// [ISL 2.0]: https://amazon-ion.github.io/ion-schema/docs/isl-2-0/spec#ieee754_float
+#[derive(Debug, Clone, PartialEq)]
+pub struct Ieee754Float {
+    value: FloatingPointNumberFormat,
+}
+
+impl Ieee754Float {
+    pub fn value(&self) -> FloatingPointNumberFormat {
+        self.value
+    }
+}
+
+impl<V: IslVersion> TypeDefinitionBuilder<V> {
+    pub fn ieee754_float(self, value: FloatingPointNumberFormat) -> Self {
+        let constraint = Ieee754Float { value };
+        self.with_constraint(constraint.into())
+    }
+}
+
+impl ValidateInternal for Ieee754Float {
+    fn validate_internal<'top: 'call, 'call, R>(
+        &'top self,
+        value: &'top IonSchemaElement,
+        ctx: &ValidationContext,
+        recorder: &'call mut R,
+    ) -> ControlFlow<()>
+    where
+        R: ViolationRecorder<'top>,
+    {
+        todo!()
+    }
+}
+
+impl WriteAsIsl<ISL_1_0> for Ieee754Float {}
+
+impl WriteAsIsl<ISL_2_0> for Ieee754Float {
+    fn write_as_isl<W: ValueWriter>(
+        &self,
+        writer: W,
+        ctx: &WriteContext<ISL_2_0>,
+    ) -> IonSchemaResult<()> {
+        todo!()
+    }
+}
+
+impl ReadConstraint<ISL_1_0> for Ieee754Float {}
+
+impl ReadConstraint<ISL_2_0> for Ieee754Float {
+    fn read_constraint(
+        ion: &Element,
+        ctx: &LoaderContext<ISL_2_0>,
+    ) -> IonSchemaResult<Option<Self>> {
+        todo!()
+    }
+}
+
+#[derive(Debug, PartialEq, Copy, Clone)]
+pub enum FloatingPointNumberFormat {
+    Binary16,
+    Binary32,
+    Binary64,
+}
+
+impl TryFrom<&Element> for FloatingPointNumberFormat {
+    type Error = IonSchemaError;
+
+    fn try_from(value: &Element) -> Result<Self, Self::Error> {
+        let text = if let Value::Symbol(s) = value.value() {
+            s.text()
+        } else {
+            None
+        };
+        use FloatingPointNumberFormat::*;
+        match text {
+            Some("binary16") => Ok(Binary16),
+            Some("binary32") => Ok(Binary32),
+            Some("binary64") => Ok(Binary64),
+            _ => invalid_schema_error(format!("Not a valid IEEE754 format symbol: {value}")),
+        }
+    }
```

Comment:
I was curious to see if the compiler would do something clever with this, like lifting out the comparison of the string prefix. To my surprise, [it did not](https://godbolt.org/z/Tvdf4sz9M).

![image](https://github.com/user-attachments/assets/4970d99f-775e-4815-8de2-90e645e3fdc2)